### PR TITLE
Add transaction support to entity updater package

### LINKED ISSUE #1905: Error running entity-updater:  index row size exceeds maximum for index "entity__public_key_natural_id"  error
**Detailed Description**
During an attempted migration using the entity-updater package we encountered a `error: index row size 2792 exceeds maximum 2712 for index "entity__public_key_natural_id" ` when addressing https://github.com/hashgraph/hedera-mirror-node/issues/509

This seems to be the result of a larger than expected key being attempted to be written.

Additionally the package does not currently revert when encountering a failure in writing.
We should wrap the writes in a transaction

**Actual Behavior**
1. Run entity-updater against mainnet with the key.csv found in https://github.com/hashgraph/hedera-mirror-node/issues/509
2. encountered error noted

**Expected Behavior**
No error on updating entity table

**Environment:**
- Java: [e.g. OpenJDK 11.0.4]
- Node: [e.g. v12.9.1-x86]
- OS: [e.g. Ubuntu 18.04]
- Version: [e.g. 1.0.1]

**Additional Context**
```shell script
2021-04-30T16:06:06.123-0000 INFO Loaded configurations from source: .../config/application.yml 
2021-04-30T16:06:06.326-0000 INFO Configure SDK client for OTHER 
2021-04-30T16:06:06.405-0000 INFO SDK client successfully configured for Operator ID 0.0.x 
2021-04-30T16:06:06.463-0000 INFO Parsing csv entity file ... 
2021-04-30T16:06:06.477-0000 INFO 4768 entities were extracted from .../keys.csv in 0 s 12.911426 ms 
2021-04-30T16:06:06.477-0000 INFO Validating entities against db and network entries ... 
2021-04-30T16:07:56.781-0000 INFO 4768 entities were retrieved and compared in 110 s 303.62668 ms, 
    1612 were found to be out-of-date 
2021-04-30T16:07:56.786-0000 DEBUG updateCriteriaCount {"auto_renew_period":210,"deleted":0,"expiration_timestamp":306,"key":1342,"memo":0,"proxy_account_id":0} 
2021-04-30T16:07:56.927-0000 DEBUG Network accountInfo calls cost 157943210 tℏ. startBalance: 2145224163379, endBalance: 2145066220169 
2021-04-30T16:07:56.928-0000 INFO Updating 1612 stale db entries with updated information ... 
(node:16262) UnhandledPromiseRejectionWarning: error: index row size 2792 exceeds maximum 2712 for index "entity__public_key_natural_id" 
    at Parser.parseErrorMessage (/home/nanaec/package/node_modules/pg-protocol/dist/parser.js:287:98) 
    at Parser.handlePacket (/home/nanaec/package/node_modules/pg-protocol/dist/parser.js:126:29) 
    at Parser.parse (/home/nanaec/package/node_modules/pg-protocol/dist/parser.js:39:38) 
    at Socket.<anonymous> (/home/nanaec/package/node_modules/pg-protocol/dist/index.js:11:42) 
    at Socket.emit (events.js:315:20) 
    at addChunk (internal/streams/readable.js:309:12) 
    at readableAddChunk (internal/streams/readable.js:284:9) 
    at Socket.Readable.push (internal/streams/readable.js:223:10) 
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23) 
(Use `node --trace-warnings ...` to show where the warning was created) 
(node:16262) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1) 
(node:16262) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code. 

```

### PR DESCRIPTION
**Pull request by Nana-EC**
**Detailed description**:
During an attempted migration using the entity-updater package we encountered a error: index row size 2792 exceeds maximum 2712 for index "entity__public_key_natural_id" when addressing https://github.com/hashgraph/hedera-mirror-node/issues/509

This was the result of a larger than expected key being attempted to be written since a non simple public_key was used.
This was an oversight in the package but a corner case handled by the importer.
Similarly the package did not utilize a transaction approach and so changes had to be manually rolled back

- Update package to ensure 1st element of KeyList contains data and not another key
- Wrap queries in transactions to ensure failures rollback any changes
- Add cache cases for database and network call to aid testing
- Add test cases to verify processing logic

**Which issue(s) this PR fixes**:
Fixes #1905 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: hedera-mirror-rest/entity-info-updater/config.js [hunk 1]
```diff
@@ -31,7 +31,8 @@ let loaded = false;
 
 const loadYaml = () => {
   try {
-    const configFile = path.join(__dirname, 'config', `application.yml`);
+    const configPath = process.env.CONFIG_PATH || `${__dirname}/config`;
```
Comment (steven-sheehy) [inline hedera-mirror-rest/entity-info-updater/config.js:34]: Probably doesn't matter for this one off, but should use `path.join`

File: hedera-mirror-rest/entity-info-updater/utils.js [hunk 2]
```diff
@@ -41,6 +42,11 @@ const longMaxValue = 9223372036854775807n;
 const realmOffset = 2n ** 32n; // realm is followed by 32 bits entity_num
 const shardOffset = 2n ** 48n; // shard is followed by 16 bits realm and 32 bits entity_num
 
+// sample location details
+const entityCachePath = process.env.ENTITY_CACHE_PATH || `${__dirname}/sample`;
```
Comment (steven-sheehy) [inline hedera-mirror-rest/entity-info-updater/utils.js:46]: path.join

File: hedera-mirror-rest/entity-info-updater/index.js [hunk 3]
```diff
@@ -62,17 +62,30 @@ const getUpdateList = async (entities) => {
...
+    logger.info(
+      `entity-info-update migration completed in ${utils.getElapsedTimeString(
+        elapsedTime
+      )} updating ${updateCount} entities`
+    );
+    return updateCount;
+  });
+};
+
+console.log(runUpdater());
```
Comment (steven-sheehy) [inline hedera-mirror-rest/entity-info-updater/index.js:87]: Use logger

File: hedera-mirror-rest/entity-info-updater/sample/keys.csv [hunk 4]
```diff
@@ -0,0 +1,4769 @@
+entity, keyAccountInfo, keyDatabase
+0.0.1000, 32240a2212200aa8e21064c61eab86e2a9c164565b4e7a9a4146106e0a6cd03a8c395a110e92, 32240a221220f9c9016c24f98132ca4a78f34d98f0393bda9431f5921367e4f834dfd2bfa4a3
```
Comment (steven-sheehy) [inline hedera-mirror-rest/entity-info-updater/sample/keys.csv:2]: This is a large file to have checked in. Is the reference input file that is used to loop over which entities to retrieve and update? If so, then it shouldn't be in a sample folder and it should be specified as the default config in `application.yaml`. I would also suggest it be gzipped.

If not, it should be removed.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]
1.2M rows added!? Please remove the cache files from the commit. They're too large, they're already out of date, they should not be checked in and there's too large of a risk they're accidentally ran in production.

Even if you force remove them by squashing/rebasing, they are forever included in our git history and now git clones will take longer unfortunately. :(

File: hedera-mirror-rest/entity-info-updater/index.js (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/index.js:87]: Solely for test purposes as a file run output. I think if I use logger I now have to parse the logger format etc

File: hedera-mirror-rest/entity-info-updater/sample/keys.csv (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/sample/keys.csv:2]: Yeah the json files map to this cdv's entity rows.
I didn't want to set this as the default as future runs might use different entity sets, so I want to ensure ether user explicitly sets the location of the csv.

I can look into gzipping and unzipping for tests

File: hedera-mirror-rest/entity-info-updater/utils.js (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/utils.js:46]: Will update

File: hedera-mirror-rest/entity-info-updater/config.js (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/config.js:34]: Will update

File: hedera-mirror-rest/entity-info-updater/index.js (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/index.js:87]: Actually I don't even need this log line anymore as I'm calling the function directly.

File: hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js [hunk 5]
```diff
@@ -99,12 +99,13 @@ const getUpdatedEntity = (dbEntity, networkEntity) => {
 
   // compare keys as buffers
   const {protoBuffer, ed25519Hex} = utils.getBufferAndEd25519HexFromKey(networkEntity.key);
-  if (Buffer.compare(dbEntity.key, protoBuffer) !== 0) {
+  const dbKey = dbEntity.key === null ? Buffer.from('') : dbEntity.key;
+  if (Buffer.compare(dbKey, protoBuffer) !== 0) {
```
Comment (xin-hedera) [inline hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js:103]: nit:

```suggestion
  const dbKey = dbEntity.key;
  if (dbKey === null || Buffer.compare(dbKey, protoBuffer) !== 0) {
```

File: hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js [hunk 6]
```diff
@@ -253,15 +257,29 @@ const getUpdateList = async (csvEntities) => {
...
   logger.info(`Updating ${entitiesToUpdate.length} stale db entries with updated information ...`);
   const updateStart = process.hrtime();
-  const updatedList = await Promise.all(entitiesToUpdate.map(dbEntityService.updateEntity));
-  const elapsedTime = process.hrtime(updateStart);
+  let updateCount = entitiesToUpdate.length;
+  try {
+    await dbEntityService.getClientConnection();
+    await dbEntityService.beginTransaction();
+    const updatedList = await Promise.all(entitiesToUpdate.map(dbEntityService.updateEntity));
+    const elapsedTime = process.hrtime(updateStart);
```
Comment (xin-hedera) [inline hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js:270]: nit, `hrtime` is deprecated

https://nodejs.org/api/process.html#process_process_hrtime_time

File: hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js [hunk 7]
```diff
@@ -253,15 +257,29 @@ const getUpdateList = async (csvEntities) => {
...
+    await dbEntityService.commitTransaction();
 
-  logger.info(`Updated ${updatedList.length} entities in ${utils.getElapsedTimeString(elapsedTime)}`);
+    logger.info(`Updated ${updatedList.length} entities in ${utils.getElapsedTimeString(elapsedTime)}`);
+  } catch (e) {
+    await dbEntityService.rollbackTransaction();
+    logger.debug(`Error updating entities in db. Rolled back updates, error: ${e}. ${e.stack}`);
+    updateCount = -1;
```
Comment (xin-hedera) [inline hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js:277]: nit, can skip setting `updateCount`, since the error is rethrown

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/entity-info-updater/dbEntityService.js [hunk 8]
```diff
@@ -103,24 +156,45 @@ const updateEntity = async (entity) => {
...
     logger.trace(`Updated entity ${entity.id}`);
   }
 };
 
+const restoreDbEntityCache = () => {
+  dbEntityCache = utils.getDbEntityCache();
+};
+
+restoreDbEntityCache();
+if (config.db.useCache === true && dbEntityCache !== {}) {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/dbEntityService.js:186]: lodash is my friend :)

File: hedera-mirror-rest/entity-info-updater/config.js (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/config.js]: Updated

File: hedera-mirror-rest/entity-info-updater/utils.js (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/utils.js]: Updated

File: hedera-mirror-rest/entity-info-updater/index.js (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/index.js]: Updated

File: hedera-mirror-rest/entity-info-updater/sample/keys.csv (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/sample/keys.csv]: Added files added gzips and added logic to gunzip all files on test run if needed

File: hedera-mirror-rest/entity-info-updater/dbEntityService.js [hunk 9]
```diff
@@ -36,6 +37,48 @@ const pool = new Pool({
   password: config.db.password,
   port: config.db.port,
 });
+let client;
+let dbEntityCache = {};
+
+const getClientConnection = async () => {
+  if (config.db.useCache === false) {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/dbEntityService.js]: Updated

File: hedera-mirror-rest/entity-info-updater/dbEntityService.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/dbEntityService.js]: Updated

File: hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js]: Updated

File: hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js]: Thanks, updated to use `process.hrtime.bigint()` instead

File: hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/entityUpdateHandler.js]: Removed

File: hedera-mirror-rest/entity-info-updater/networkEntityService.js [hunk 10]
```diff
@@ -105,11 +117,18 @@ if (!clientConfigured) {
     );
   }
 
-  operatorId = AccountId.fromString(config.sdkClient.operatorId);
-  const operatorKey = PrivateKey.fromString(config.sdkClient.operatorKey);
-  client.setOperator(operatorId, operatorKey);
-  clientConfigured = true;
-  logger.info(`SDK client successfully configured for Operator ID ${operatorId}`);
+  restoreNetworkEntityCache();
+  if (config.sdkClient.useCache === true && networkEntityCache !== {}) {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/networkEntityService.js]: Updated

File: hedera-mirror-rest/entity-info-updater/networkEntityService.js [hunk 11]
```diff
@@ -42,9 +44,15 @@ const getAccountInfo = async (accountId) => {
   logger.trace(`Retrieve account info for ${accountId}`);
   let accountInfo;
   try {
-    accountInfo = await new AccountInfoQuery().setAccountId(accountId).execute(client);
+    if (config.sdkClient.useCache && networkEntityCache[accountId.toString()] !== undefined) {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/networkEntityService.js]: Updated

File: hedera-mirror-rest/entity-info-updater/networkEntityService.js [hunk 12]
```diff
@@ -42,9 +44,15 @@ const getAccountInfo = async (accountId) => {
...
+      accountInfo = AccountInfo.fromBytes(Buffer.from(networkEntityCache[accountId.toString()]));
+      logger.trace(`Retrieved ${accountId.toString()} from cache`);
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-rest/entity-info-updater/networkEntityService.js]: Updated

**Review by ijungmann — APPROVED** [review]

**Review by xin-hedera — APPROVED** [review]
LGTM

**Comment by steven-sheehy** [comment]
already reviewed...

**Pull request merged**